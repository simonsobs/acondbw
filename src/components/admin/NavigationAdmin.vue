<template>
  <div class="navigation" style="position: relative">
    <v-list shaped>
      <v-list-item
        link
        router
        v-for="(link, index) in links"
        :key="index"
        :to="link.to"
      >
        <v-list-item-action>
          <v-icon v-text="link.icon"></v-icon>
        </v-list-item-action>
        <v-list-item-content>
          <v-list-item-title
            v-text="link.title"
            class="capitalize"
          ></v-list-item-title>
        </v-list-item-content>
      </v-list-item>
    </v-list>
  </div>
</template>

<script>
export default {
  name: "NavigationAdmin",
  data: () => ({
    edges: null,
    links: [
      {
        title: "versions",
        to: { name: "AdminVersion" },
        icon: "mdi-scatter-plot-outline",
      },
      {
        title: "log",
        to: { name: "AdminLog" },
        icon: "mdi-exclamation-thick",
      },
      {
        title: "config",
        to: { name: "AdminConfig" },
        icon: "mdi-cog",
      },
      {
        title: "theme",
        to: { name: "AdminTheme" },
        icon: "mdi-format-color-fill",
      },
      {
        title: "scratch",
        to: { name: "AdminScratch" },
        icon: "mdi-note-edit-outline",
      },
    ],
  }),
};
</script>

<style scoped>
.capitalize {
  text-transform: capitalize;
}
</style>
