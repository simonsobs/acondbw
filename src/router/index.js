import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
