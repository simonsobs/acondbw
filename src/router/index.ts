import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
