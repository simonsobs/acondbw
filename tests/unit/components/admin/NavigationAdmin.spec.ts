import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { shallowMount, createLocalVue } from "@vue/test-utils";

import NavigationAdmin from "@/components/admin/NavigationAdmin.vue";
import { createRouter } from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("App.vue", () => {
  let localVue: ReturnType<typeof createLocalVue>;
  let vuetify: Vuetify;
  let router: ReturnType<typeof createRouter>;
  let wrapper: ReturnType<typeof shallowMount>;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    router = createRouter();
    wrapper = shallowMount(NavigationAdmin, {
      localVue,
      vuetify,
      router,
    });
  });

  it("match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
