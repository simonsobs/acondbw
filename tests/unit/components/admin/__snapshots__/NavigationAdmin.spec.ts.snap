// Vitest Snapshot v1

exports[`App.vue > match snapshot 1`] = `
"<div data-v-dcc4e5c8=\\"\\" class=\\"navigation\\" style=\\"position: relative;\\">
  <vuecomponent-stub data-v-dcc4e5c8=\\"\\" shaped=\\"true\\" tag=\\"div\\">
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-scatter-plot-outline</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">versions</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-exclamation-thick</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">log</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-cog</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">config</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-format-color-fill</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">theme</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-account-multiple</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">users</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <vuecomponent-stub data-v-dcc4e5c8=\\"\\" activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\">mdi-shape</vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\">
        <vuecomponent-stub data-v-dcc4e5c8=\\"\\" tag=\\"div\\" class=\\"capitalize condensed-font\\">product types</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
  </vuecomponent-stub>
</div>"
`;
